common/logger/src/lib.rs

```diff
@@ -1,74 +1,531 @@
 #[macro_export]
 macro_rules! info {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Info) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::info!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define trace macro that specify libra as the target
 #[macro_export]
 macro_rules! trace {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Trace) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::trace!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
 }
 
 /// Define warn macro that specify libra as the target
 #[macro_export]
 macro_rules! warn {
-    ($($arg:tt)+) => (
+    ($($arg:tt)+) => ({
+        if $crate::struct_log_enabled!($crate::log::Level::Warn) {
+            $crate::struct_log!($($arg)+);
+        }
         $crate::log::warn!(target: $crate::DEFAULT_TARGET, $($arg)+);
-    )
+    })
+}
+
+#[macro_export]
+#[cfg(feature = "no_struct_log")]
+macro_rules! struct_log_enabled {
+    ($level:expr) => {
+        false
+    };
+}
+
+#[macro_export]
+#[cfg(not(feature = "no_struct_log"))]
+macro_rules! struct_log_enabled {
+    ($level:expr) => {
+        $crate::log::log_enabled!(target: $crate::DEFAULT_TARGET, $level)
+            && $crate::struct_logger_enabled()
+    };
+}
+
+#[macro_export]
+#[cfg(feature = "no_struct_log")]
+macro_rules! struct_log {
+    ($($arg:tt)+) => {};
+}
+
+#[macro_export]
+#[cfg(not(feature = "no_struct_log"))]
+macro_rules! struct_log {
+    ($($arg:tt)+) => {
+        let mut entry = $crate::StructuredLogEntry::new_unnamed();
+        $crate::format_struct_args_and_pattern!(entry, $($arg)+);
+        $crate::send_struct_log!(entry);
+    }
+}
+
+#[macro_export]
+macro_rules! send_struct_log {
+    ($entry:expr) => {
+        if $crate::struct_logger_enabled() {
+            let mut entry = $entry;
+            entry.timestamp($crate::Utc::now().format("%F %T").to_string());
+            entry.module(module_path!());
+            entry.location($crate::location!());
+            entry.git_rev($crate::git_rev!());
+            entry.send();
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! location {
+    () => {
+        concat!(file!(), ":", line!())
+    };
+}
+
+// GIT_REV env need to be set during _compile_ time to have this var populated
+#[macro_export]
+macro_rules! git_rev {
+    () => {
+        option_env!("GIT_REV")
+    };
+}
+
+#[macro_export]
+macro_rules! format_struct_args_and_pattern {
+    ($entry:ident, $fmt:expr) => {
+        $entry.pattern($fmt);
+    };
+    ($entry:ident, $fmt:expr,) => {
+        $entry.pattern($fmt);
+    };
+    ($entry:ident, $fmt:expr, $($arg:tt)+) => {
+        $entry.pattern($fmt);
+        $crate::format_struct_args!($entry, 0, $($arg)+);
+    }
+}
+
+#[macro_export]
+macro_rules! format_struct_args {
```

Comment:
This seems a lot more complicated then what they did in log's structured log case:
```
(target: $target:expr, $lvl:expr, ($($arg:expr),*) { $($key:ident : $value:expr),* }) => {{
        let lvl = log::Level::Info;
        if lvl <= $crate::STATIC_MAX_LEVEL && lvl <= $crate::max_level() {
            $crate::__private_api_log(
                __log_format_args!($($arg),*),
                lvl,
                &(__log_module_path!(), __log_module_path!(), __log_file!(), __log_line!()),
                Some(&[$((__log_stringify!($key), &$value)),*])
```
I guess its because of the flexibility you were trying to obtain?

Reply:
Logging macro piggy backs on macro from `fmt`, while we need to do formatting ourselves, thats why it is more complex

Reply:
We should have a macro competition to see if someone can minimize this while still maintaining readability.